onnxruntime/core/providers/cpu/controlflow/scan_9.cc

```diff
@@ -0,0 +1,470 @@
+        .TypeConstraint("V", OpSchema::all_tensor_types(), "All Tensor types")
+}
+*/
+
+class ScanImpl {
+ public:
+  ScanImpl(OpKernelContextInternal& context,
+           const SessionState& session_state,
+           int64_t num_scan_inputs,
+           const std::vector<int64_t>& input_directions,
+           const std::vector<int64_t>& output_directions,
+           const std::vector<int64_t>& axes);
+
+  // Initialize by validating all the inputs, and allocating the output tensors
+  Status Initialize();
+
+  // Execute the batch, by iterating the sequence in each batch entry
+  // and calling the subgraph with each item in the sequence.
+  Status Execute();
+
+ private:
+  // validate inputs and setup batch size and max sequence length.
+  Status ValidateInput();
+
+  Status ValidateSubgraphInput(int start_input, int end_input,
+                               const std::vector<const NodeArg*>& graph_inputs);
+
+  // setup inputs to subgraph, transposing if necessary
+  Status SetupInputs();
+
+  Status AllocateOutputTensors();
+  Status CreateLoopStateVariables(std::vector<LoopStateVariable>& loop_state_variables);
+
+  using ConstTensorSlicerIterators = std::vector<MLValueTensorSlicer<const MLValue>::Iterator>;
+  using MutableTensorSlicerIterators = std::vector<MLValueTensorSlicer<MLValue>::Iterator>;
+
+  OpKernelContextInternal& context_;
+  const SessionState& session_state_;
+  const GraphViewer& subgraph_;
+
+  int num_loop_state_variables_;
+  int num_scan_inputs_;
+  int num_variadic_inputs_;
+  int num_variadic_outputs_;
+
+  int64_t sequence_len_ = -1;
+
+  const std::vector<int64_t>& input_directions_;
+  const std::vector<int64_t>& output_directions_;
+  const std::vector<int64_t>& axes_from_attribute_;
+  std::vector<int64_t> axes_;
+
+  // inputs for graph. either original input value or transposed input if an axis other than 0 was specified
+  std::vector<MLValue> inputs_;
+
+  std::vector<std::string> subgraph_output_names_;
+  std::vector<std::unique_ptr<OutputIterator>> output_iterators_;
+
+  std::unordered_map<std::string, const MLValue*> implicit_inputs_;
+};
+
+template <>
+Scan<9>::Scan(const OpKernelInfo& info) : OpKernel(info) {
+  // make sure the attribute was present even though we don't need it here.
+  // The GraphProto is loaded as a Graph instance by main Graph::Resolve,
+  // and a SessionState instance for executing the subgraph is created by InferenceSession.
+  // This is available via Info().GetSubgraphSessionState("attribute_name") when Compute is called.
+  ONNX_NAMESPACE::GraphProto proto;
+  ORT_ENFORCE(info.GetAttr<ONNX_NAMESPACE::GraphProto>("body", &proto).IsOK());
+  (void)proto;
+
+  ORT_ENFORCE(info.GetAttr<int64_t>("num_scan_inputs", &num_scan_inputs_).IsOK());
+
+  ReadDirections(info, "scan_input_directions", input_directions_, num_scan_inputs_);
+
+  // we don't know the concrete number of outputs until Compute is called, as the subgraph may have optional inputs
```

Comment:
I don't understand this comment. But I guess this relates to the offline discussion we have been having? I assume this is not critical here anyway. But I believe all this information is available at kernel instantiation time.

Reply:
You're right - the information in available. Will update accordingly.